README.md
# Filer (v5.0.0)
A 'cloud' file-explorer in NodeJS that utilizes the disk-space on the host's machine to transfer or download files from the host machine.

# Contents
-	Changelog(s) can be found [here](CHANGELOG.md).
-	Tutorial can be found [here](TUTORIAL.md).
-	Installation instruction(s) can be found [here](INSTALLATION.md).

# Note
-   When using this program, Please note that the connecting device must be the on the same network as the host device. (Host is the one running this program!)

# Why make this?
I wanted to create my own knock-off 'cloud' storage. So I created this! Any file placed within the root folder specified will be fully accesible to other devices on the same local network. For example: if I uploaded a .png file to the root folder. My phone can access the item if its on the same network connection as the host (my computer).

CHANGELOG.md
# v5.0.0
-	Updated the UI with new visual elements.
-	Removed the confusing welcome page, opening a directory or downloading a file? just click the directory/file name. (Updated endpoint, both directory and files both go throuh `/open`)
-	More reliable check if item inside a directory is a directory.
-	More stable and reliable system thanks to Promises (Last time, all responses have a 200ms buffer before getting the response. Now it happenc async-ly)
-	Replaced `/open_file` with `/open` and it takes two query parameter: `path` and `type`. `path` is the current path, `type` is either `directory` or `file`, this helps the endpoint extinguished whether to download or open the directory.
-	Due to the `/open` endpoint having the `path` parameter, it means that anyone can place a valid directory and snoop around!. This is quickly countered by a check if the current path includes the `argument` passed in when running the server!
-	Removed unreliable `count` variable that "helps". (Previous changelog: `Now recursive file or directory viewing is enabled through the help of a counter variable.`).
-	Now, there is a `../` directory, it brings you a step back (in terms of directory)
-	Added a tutorial. (md file)

# v4.0.1
-	Added partials for the EJS files (more consistent changes server-wide for the front-end)
-	The page will display the total number of items found in the current directory (was present in v3).
-	Now, when running the server, a argument is needed to be passed which is the root folder of the server
	-	So, instead of `node filer.js`, it is now `node filer.js ./` or maybe `node filer.js ./files` considering there is a directory called `files` within the server  folder.
-	Updated the welcome page
-	Every page will have a footer anchor tag which displays `View the welcome page`. The welcome page includes tutorial
-   Removed the dotenv dependency

## Bugs:
-	Some files (when trying to open) may download the file instead. (Persistent bug!)
-	Some files, when downloading will give a 'NotFound' error!

# v4.0.0
-	Instead of relying on response.write() for writing HTML. The site now uses EJS Templating for the front-end. (More reliable front-end solution)
-	Removed /test endpoint
-	Added a backend API. New endpoint: `/api/directory?d=(directory)`. This endpoint will return a JSON response containing the directory contents.
-	New layout (front-end)
	-	`[X]` - (Attempts to) Open the file in the browser
	-	`[FILE NAME]` - Open the directory (opening a directory on a file will return an error)
	-	`[↓]` - Downloads the file (with the correct name)
-	Deleted welcome.txt as the /welcome endpoint now serves a proper render of the welcome ejs file
-	Instead of all the EJS files being inside views/pages, it is now stored in the views folder
-	Updated the index page.

# v4.0.0-internal_beta
-	Added EJS Templating. (More reliable loading times)

# v3.3.0
-	Directories with empty contents will no longer crash the server.

# v3.2.2
-   Added support for hosting on servers such as c9 and heroku.
-   Renamed Filer.js to filer.js
-   Removed welcome.txt
-   Introduced the /welcome endpoint (still in the making!)
-   Updated dependencies (added dotenv) for writing custom environment variables (in hopes of using it in the future!)

# v3.2.1
-	Better logging! (Logged correct things at the right time)
-   Replaced `network` dependency with `ip` so that android is supported!	-	Removed a logging statement that logged the Private IP Address that was used in debugging!
-	Removed the version in the html head title. Instead of `Filer v3.0.1`, now is `Filer`

# v3.2.0
-	Added support for android!
-   Replaced `network` dependency with `ip` so that android is supported!
-	Removed the argument that is needed to be given when running the program. So, instead of: `node index.js ./`, its `node index.js` and the default folder that is hosted is the root folder where the index.js file is

# v3.0.1
-   Changes are made to accomodate folders with a lot of files inside.
-	Visiting the / endpoint will redirect you to the directory endpoint pointing to the root folder.	    -   Response is only sent back after 200ms as the server need to get the directory contents first!

# v3.0.0
-   More verbose logging at the console.
-	Visiting the / endpoint will redirect you to the directory endpoint pointing to the root folder.
-	Users can select to OPEN the file (By clicking on a new anchor tag which says the file name or download by clicking the anchor tag which shows the filetype
-	There is now a header (h1) stating the current directory you are in
-	Added temporary styling (font and colours)
-	Now, directory names will display with `./` at the front
-	Removed /info endpoint.
-	Now files are downloaded to the device upon clicking the anchor tag. (Instead of Opening it).
-	Now instead of console.log-ing everything, each log text is pushed to an array which is later logged out by joining the items in the array.
-	Added 'ROOT' directory anchor tag
-	The server will be over HTTP and there is no plans of implementing HTTPS. Implementing HTTPS is up to the user.
-   Now, downloading the file will download with its file name instead of its endpoint `welcome.txt` instead of `open_file.txt`

# v2.2.0
-   Makeshift User Interface for users to interact with the server.
-	To run, now the user must pass in an argument of the root folder for the server. For example: `node index.js ./personal/`
-	Updated `/info` endpoint with correct display of information
-	Now, the `type` of item is an anchor tag which when clicked is the same as clicking the endpoint link to go to the directory or open the file [UI].
-   Server still runs on HTTP! Please do not store sensitive information within your root folder.

# v2.1.0
-	Now, items are supposed to be placed within the 'storage' folder
-   Changed default port from 3000 to 3030. (Will randomise this in the future)	-   Removed Feature: The first index of the array is now a link which directs to the `root` folder.

Filer.js
let express = require('express');
let fs = require('fs');
let app = express();

const IP_ADDRESS = require('ip').address()
const PORT = 3030;

// The IP and PORT is stored here
let server_on = `${IP_ADDRESS}:${PORT}`

// Current directory contents
let __dir_contents__ = [];
let dir_log = ["/"]
let count = 0;

// Log text output
let log_text = [];

// It will take the project folder and search inside "storage" dir

console.log(`From Directory: ${__dirname}`)

app.use((request, response, next) => {

    // Logging!
    log_text.push(`GET: ${request.url}`)
    log_text.push(`Query (d): ${request.query.d}`)
    log_text.push(`Is req url == /?:  ${request.url === "/"}`)
    log_text.push(`Directory log: ${JSON.stringify(dir_log)}`)
    log_text.push(`Current Directory: ${JSON.stringify(dir_log[count])}`)
    log_text.push(`Step Back Directory: ${JSON.stringify(dir_log[count-1])}`)

    // Count is increase because every directory that the user visits is pushed to dir_log, so if the user visited that page for the 39th time, so its current directory is dir_log[count]
    // dir_log[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
    count++

    if (request.url === "/") {
        dir_log.push("/")
    } else {
        dir_log.push(request.query.d)
    }

    next();
})

app.get('/', (request, response) => {
    // Redirect to the directory endpoint with query parameter d pointing to root folder
    response.redirect("/directory?d=./")
})

app.get('/directory', (request, response) => {
    response.writeHead(200, {
        'Content-Type': 'text/html'
    });

    if (request.query.d == "./") {
        __getcontents__(__dirname, dir_log)
    } else {
        __getcontents__(__dirname + "/" + request.query.d, dir_log);
    }

    response.write(`
        <head>

            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Filer</title>
            <style>

                @import url('https://fonts.googleapis.com/css?family=Roboto+Mono&display=swap');

                body{
                    color: black;
                    font-family: 'Roboto Mono', monospace;
                }

                a{
                    color:black;
                    background: yellow;
                }

                h1,h2,h3,h4,h5,h6{
                    font-weight: 400;
                }
            </style>
        </head>`)

    response.write(`<h1>${dir_log[count]}</h1>`)
    response.write(`[<a href="/">DIR</a>] /<br>`);

    setTimeout(() => {
        for (let i = 0; i < __dir_contents__.length; i++) {

            // If the item is a dir:
            // TEMP REMOVE
            if (__dir_contents__[i].type == "dir") {
                response.write(`[<a href="${__dir_contents__[i].download_html_link}">${__dir_contents__[i].type.toUpperCase()}</a>] ${__dir_contents__[i].name}<br>`);
            } else {
                response.write(`[<a href="${__dir_contents__[i].download_html_link}">${__dir_contents__[i].type.toUpperCase()}</a>] <a href="${__dir_contents__[i].open_html_link}">${__dir_contents__[i].name}</a><br>`);
            }
        }

        // End the response!
        response.end(`${__dir_contents__.length} item(s) found in this directory`);
    }, 200);
})

app.get('/open_file', (request, response) => {
    let t = request.query.t.toUpperCase();

    if (t == "OPEN") {
        response.sendFile(__dirname + "/" + request.query.f)
    } else {
        response.download(__dirname + "/" + request.query.f, request.query.f)
    }
})
// Listen on the private ip and PORT
app.listen(PORT, () => {
    console.log(`Listening on: ${server_on} `);
    __getcontents__(__dirname, dir_log)
})

// Function to get contents of the directory
function __getcontents__(path, input_dir_history) {

    // Empty the dir_contents
    __dir_contents__ = [];



    // Read the directory for both files and directories
    fs.readdir(path, (err, items) => {

        log_text.push(`Step Back Redirect Link: http://${server_on}/directory?d=${input_dir_history[count-1]}`)
        log_text.push(`Length of item inside the current directory: ${items.length}`)
        log_text.push(`\n ==============================================`)

        // If the items found inside is not empty ['file1', 'file2']
        if (items.length > 0) {

            // Loop through the items
            for (let i = 0; i < items.length; i++) {

                // C:\Users\nabil\Documents\mailpushserver/storage//personal/images/visualres2.png
                // After /storage/ is inpput_dir_history
                // The last / is items[i]
                let inner_dir = __dirname + input_dir_history[count] + "/" + items[i];

                // Check if the items inside are files or direcotyr
                fs.readdir(inner_dir, (err, inner_dir_test) => {

                    // Not a dir items return undefined!
                    if (!inner_dir_test) {

                        let filetype = items[i].split(".")[items[i].split(".").length - 1]

                        // NOT A DIR
                        __dir_contents__[i] = {
                            "name": items[i],
                            "type": filetype,
                            "full_path": inner_dir,
                            "download_html_link": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=download`,
                            "open_html_link": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=open`,
                            "link": {
                                "dir": null,
                                "file": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=download`,
                                "open": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=open`
                            },
                        }

                        // as you see above, input_dir_history[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
                    } else {

                        // IS A DIR
                        __dir_contents__[i] = {
                            "name": `./${items[i]}`,
                            "type": "dir",
                            "full_path": inner_dir,
                            "download_html_link": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                            "open_html_link": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                            "link": {
                                "dir": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                                "file": null
                            },
                        }
                        // as you see above, input_dir_history[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
                    }
                })
            }
        }
    })
}

filer.js
const express = require('express');
const fs = require('fs');
const app = express();

// The IP and PORT is stored here
let server_on = "";

// Current directory contents
let __fromdir = process.argv[2]
// set the view engine to ejs
app.set('view engine', 'ejs');
console.log(`From Directory: ${__fromdir}`)

app.get('/', (request, response) => {
    // Redirect to the directory endpoint with query parameter d pointing to root folder
    response.redirect(`/open?path=${__fromdir}&type=directory`)
})

// To get a file
app.get('/open', (request, response) => {

    let path = request.query.path;
    let type = request.query.type;

    if (path.includes(process.argv[2]) === false) {
        response.render("forbidden")
    } else {
        if (type == "directory") {
            __getcontents__(path)
                .then(d => {
                    response.render("index", {
                        __dircontents: d,
                        path: path,
                        step_back_dir: path.split("/").slice(0, path.split("/").length - 1).join("/"),
                        initial_path: __fromdir
                    })
                })
        }
    }

    if (type == "file") {
        response.download(path, (err) => {
            if (err) console.log(err)
        })
    }
})

// Function to get contents of the directory
let __getcontents__ = (path) => {
    return new Promise((resolve, reject) => {

        let contents = []

        // Empty the dir_contents
        // Read the directory for both files and directories
        fs.readdir(path, (err, items) => {
            if (err) console.log(err)

            if (items.length !== 0) {
                items.forEach((item, n) => {
                    get_html_link(path + "/" + item)
                        .then(link => {

                            if (link.includes("type=directory")) {
                                contents.push({
                                    "name": `/${item}`,
                                    "open_html_link": link,
                                    "type": "directory"
                                })
                            } else {
                                contents.push({
                                    "name": `${item}`,
                                    "open_html_link": link,
                                    "type": "file"
                                })
                            }

                            if (contents.length == items.length) resolve(contents)

                        })
                })
            } else {
                resolve(contents)
            }

        })
    })
}

let get_html_link = (path) => {
    return new Promise((resolve, reject) => {
        fs.readdir(path, (err, files) => {
            if (files !== undefined) {
                resolve(`/open?path=${path}&type=directory`)
            } else {
                resolve(`/open?path=${path}&type=file`)
            }
        })
    })
}

// MAIN INITIALISER
if (!process.env.IP || !process.env.PORT) {
    // Listen on the PRIVATE IP and PORT 3030
    app.listen(3030, require('ip').address(), () => {
        server_on = `${require('ip').address()}:3030`
        console.log(`HOSTING LOCALLY: Listening on: http://${server_on} `);

        __getcontents__(__fromdir)
    })
} else {
    // Listen on the process.env.IP and PORT
    server_on = `${process.env.IP}:${process.env.PORT}`
    app.listen(process.env.PORT, () => {
        console.log(`HOSTING OTHER SERVER: Listening on: http://${server_on} `);

        __getcontents__(__fromdir)
    })
}

index.js
let express = require('express');
let fs = require('fs');
let app = express();

const IP_ADDRESS = require('ip').address()
const PORT = 3030;

// The IP and PORT is stored here
let server_on = `${IP_ADDRESS}:${PORT}`

// Current directory contents
let __dir_contents__ = [];
let dir_log = ["/"]
let count = 0;

// Log text output
let log_text = [];

// It will take the project folder and search inside "storage" dir

console.log(`From Directory: ${__dirname}`)

app.use((request, response, next) => {

    // Logging!
    log_text.push(`GET: ${request.url}`)
    log_text.push(`Query (d): ${request.query.d}`)
    log_text.push(`Is req url == /?:  ${request.url === "/"}`)
    log_text.push(`Directory log: ${JSON.stringify(dir_log)}`)
    log_text.push(`Current Directory: ${JSON.stringify(dir_log[count])}`)
    log_text.push(`Step Back Directory: ${JSON.stringify(dir_log[count-1])}`)

    // Count is increase because every directory that the user visits is pushed to dir_log, so if the user visited that page for the 39th time, so its current directory is dir_log[count]
    // dir_log[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
    count++

    if (request.url === "/") {
        dir_log.push("/")
    } else {
        dir_log.push(request.query.d)
    }

    next();
})

app.get('/', (request, response) => {
    // Redirect to the directory endpoint with query parameter d pointing to root folder
    response.redirect("/directory?d=./")
})

app.get('/directory', (request, response) => {
    response.writeHead(200, {
        'Content-Type': 'text/html'
    });

    if (request.query.d == "./") {
        __getcontents__(__dirname, dir_log)
    } else {
        __getcontents__(__dirname + "/" + request.query.d, dir_log);
    }

    response.write(`
        <head>

            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>Filer v3.0.1</title>
            <style>

                @import url('https://fonts.googleapis.com/css?family=Roboto+Mono&display=swap');

                body{
                    color: black;
                    font-family: 'Roboto Mono', monospace;
                }

                a{
                    color:black;
                    background: yellow;
                }

                h1,h2,h3,h4,h5,h6{
                    font-weight: 400;
                }
            </style>
        </head>`)

    response.write(`<h1>${dir_log[count]}</h1>`)
    response.write(`[<a href="/">DIR</a>] /<br>`);

    setTimeout(() => {
        for (let i = 0; i < __dir_contents__.length; i++) {

            // If the item is a dir:
            // TEMP REMOVE
            if (__dir_contents__[i].type == "dir") {
                response.write(`[<a href="${__dir_contents__[i].download_html_link}">${__dir_contents__[i].type.toUpperCase()}</a>] ${__dir_contents__[i].name}<br>`);
            } else {
                response.write(`[<a href="${__dir_contents__[i].download_html_link}">${__dir_contents__[i].type.toUpperCase()}</a>] <a href="${__dir_contents__[i].open_html_link}">${__dir_contents__[i].name}</a><br>`);
            }
        }

        // End the response!
        response.end(`${__dir_contents__.length} item(s) found in this directory`);
    }, 200);
})

app.get('/open_file', (request, response) => {
    let t = request.query.t.toUpperCase();

    if (t == "OPEN") {
        response.sendFile(__dirname + "/" + request.query.f)
    } else {
        response.download(__dirname + "/" + request.query.f, request.query.f)
    }
})
// Listen on the private ip and PORT
app.listen(PORT, () => {
    console.log(`Listening on: ${server_on} `);
    __getcontents__(__dirname, dir_log)
})

// Function to get contents of the directory
function __getcontents__(path, input_dir_history) {

    // Empty the dir_contents
    __dir_contents__ = [];



    // Read the directory for both files and directories
    fs.readdir(path, (err, items) => {

        log_text.push(`Step Back Redirect Link: http://${server_on}/directory?d=${input_dir_history[count-1]}`)
        log_text.push(`Length of item inside the current directory: ${items.length}`)
        log_text.push(`\n ==============================================`)

        // If the items found inside is not empty ['file1', 'file2']
        if (items.length > 0) {

            // Loop through the items
            for (let i = 0; i < items.length; i++) {

                // C:\Users\nabil\Documents\mailpushserver/storage//personal/images/visualres2.png
                // After /storage/ is inpput_dir_history
                // The last / is items[i]
                let inner_dir = __dirname + input_dir_history[count] + "/" + items[i];

                // Check if the items inside are files or direcotyr
                fs.readdir(inner_dir, (err, inner_dir_test) => {

                    // Not a dir items return undefined!
                    if (!inner_dir_test) {

                        let filetype = items[i].split(".")[items[i].split(".").length - 1]

                        // NOT A DIR
                        __dir_contents__[i] = {
                            "name": items[i],
                            "type": filetype,
                            "full_path": inner_dir,
                            "download_html_link": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=download`,
                            "open_html_link": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=open`,
                            "link": {
                                "dir": null,
                                "file": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=download`,
                                "open": `http://${server_on}/open_file?f=${input_dir_history[count]}/${items[i]}&t=open`
                            },
                        }

                        // as you see above, input_dir_history[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
                    } else {

                        // IS A DIR
                        __dir_contents__[i] = {
                            "name": `./${items[i]}`,
                            "type": "dir",
                            "full_path": inner_dir,
                            "download_html_link": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                            "open_html_link": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                            "link": {
                                "dir": `http://${server_on}/directory?d=${input_dir_history[count]}/${items[i]}`,
                                "file": null
                            },
                        }
                        // as you see above, input_dir_history[count] is to get the current latest directory, the directory that is pushed is all including its parent(s)
                    }
                })
            }
        }
    })
}